lib/logging/lib/logging.c

```diff
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2015-2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <syslog.h>
+
+#include <hse_util/event_counter.h>
+#include <hse_util/minmax.h>
+#include <hse_util/mutex.h>
+#include <hse_ikvdb/hse_gparams.h>
+
+#include <logging/logging.h>
+#include <logging/types.h>
+
+_Thread_local char scratch_buffer[4096];
+static bool initialized;
+static FILE *log_file = NULL;
+
+merr_t
+logging_init(void)
+{
+    FILE *fp = NULL;
+
+    if (!hse_gparams.gp_logging.enabled)
+        return 0;
+
+    if (hse_gparams.gp_logging.destination == LD_STDOUT) {
+        fp = stdout;
+    } else if (hse_gparams.gp_logging.destination == LD_STDERR) {
+        fp = stderr;
+    } else if (hse_gparams.gp_logging.destination == LD_FILE) {
+        fp = fopen(hse_gparams.gp_logging.path, "a");
+        if (!fp) {
+            const merr_t err = merr(errno);
+            log_errx("failed to open log file %s",
+                     err, hse_gparams.gp_logging.path);
+            return err;
+        }
+
+        setlinebuf(fp);
+    }
+
+    if (!initialized) {
+        log_file = fp;
+        initialized = true;
+    }
+
+    return 0;
+}
+
+void
+logging_fini(void)
+{
+    if (log_file && log_file != stdout && log_file != stderr)
+        fclose(log_file);
+    log_file = NULL;
+
+    initialized = false;
+}
+
+void
+log_impl(
+    struct event_counter *ev, /* contains call site info and pri     */
+    merr_t err,               /* error value                         */
+    const char *fmt,          /* the platform-specific format string */
+    ...)                      /* variable-length argument list       */
+{
+    int rc;
+    va_list args;
+    uint64_t now;
+
+    if (ev->ev_pri > hse_gparams.gp_logging.level || !hse_gparams.gp_logging.enabled)
+        return;
+
+    event_counter(ev);
+
+    now = get_time_ns();
+    if (now < ev->ev_priv)
+        return;
+
+    ev->ev_priv = now + hse_gparams.gp_logging.squelch_ns;
+
+    if (err) {
+        char buf[256];
+
+        merr_strerror(err, buf, sizeof(buf));
+
+        rc = snprintf(scratch_buffer, sizeof(scratch_buffer), "[HSE] %s:%d %s: %s: %s", ev->ev_file,
```

Comment:
Note that when the error is not 0 it will look like:

```
Aug 18 19:32:43 campise kmt[1259132]: [HSE] lib/config/argv.c:101 argv_deserialize_to_params: Applying KVS runtime param cn_maint_disable from paramv: <strerror(merr_errno(err))>
```

which seemed to be the most common placement of `@@e`

Reply:
Agreed, the error message shuld always be at the end!